### PR TITLE
Relation analysis with immutable and lazy data structures

### PR DESCRIPTION
**Pull request by natgavrilenko**

### CODE REVIEW

File: dartagnan/src/test/java/com/dat3m/dartagnan/encoding/RelationAnalysisTest.java [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+    private List<String> listFiles(Path path) throws IOException {
+        List<String> result = new LinkedList<>();
+        try (DirectoryStream<Path> files = Files.newDirectoryStream(path)) {
+            for (Path file : files) {
+                if (Files.isDirectory(file)) {
+                    result.addAll(listFiles(file.toAbsolutePath()));
+                } else {
+                    String filePath = file.toAbsolutePath().toString();
+                    if (filePath.endsWith(".litmus") || filePath.endsWith(".ll") ||
+                            filePath.endsWith(".c") || filePath.endsWith(".spv.dis")) {
```
Comment (hernanponcedeleon) [inline dartagnan/src/test/java/com/dat3m/dartagnan/encoding/RelationAnalysisTest.java:117]: The CI never really runs *.c test, thus this condition is useless

File: dartagnan/src/main/java/com/dat3m/dartagnan/wmm/analysis/LazyRelationAnalysis.java [hunk 2]
```diff
@@ -0,0 +1,651 @@
...
+            long start = System.currentTimeMillis();
+            EventGraph must = computeInternalDependencies(EnumSet.of(ADDR));
+            time(definition, start, System.currentTimeMillis());
+            return new RelationAnalysis.Knowledge(must, must);
+        }
+
+        @Override
+        public RelationAnalysis.Knowledge visitInternalDataDependency(DirectDataDependency definition) {
+            long start = System.currentTimeMillis();
+            EventGraph must = computeInternalDependencies(EnumSet.of(DATA, CTRL, OTHER));
```
[the remark this answers is not in the document]
Comment (ThomasHaas) [reply dartagnan/src/main/java/com/dat3m/dartagnan/wmm/analysis/LazyRelationAnalysis.java:232]: The reason is that if a register is used to determine control-flow in e.g. a jump, the jump itself is still data-dependent on the register but only subsequent instructions are control-dependent on it.
```
r = ...;
if (r) goto ... // Data-dependent on r and thus idd-edge exists
   x = 42; // Control-dependent on r, no idd-edge
```

File: dartagnan/src/main/java/com/dat3m/dartagnan/wmm/analysis/LazyRelationAnalysis.java [hunk 3]
```diff
@@ -0,0 +1,651 @@
...
+
+        @Override
+        public RelationAnalysis.Knowledge visitControlDependency(DirectControlDependency definition) {
+            long start = System.currentTimeMillis();
+            Map<Event, Set<Event>> data = new HashMap<>();
+            program.getThreads().forEach(t -> t.getEvents(CondJump.class).forEach(j -> {
+                if (!j.isGoto() && !j.isDead()) {
+                    ImmutableSet<Event> range = ((j instanceof IfAsJump ifJump)
+                            ? ifJump.getBranchesEvents()
+                            : j.getSuccessor().getSuccessors()).stream()
```
[the remark this answers is not in the document]
Comment (ThomasHaas) [reply dartagnan/src/main/java/com/dat3m/dartagnan/wmm/analysis/LazyRelationAnalysis.java:208]: IIRC, `getSuccessors()` also returns the event itself, so this code actually only skips `j`.
I think we should rename that function :)

File: dartagnan/src/main/java/com/dat3m/dartagnan/wmm/analysis/LazyRelationAnalysis.java [hunk 4]
```diff
@@ -0,0 +1,651 @@
...
+            Set<Event> range = visibleEvents.stream().filter(e2 -> visibleEvents.stream()
+                            .anyMatch(e1 -> e1.getThread().equals(e2.getThread())
+                                    && e1.getGlobalId() < e2.getGlobalId()
+                                    && !exec.areMutuallyExclusive(e1, e2)))
+                    .collect(toSet());
+            EventGraph must = new LazyEventGraph(domain, range,
+                    (e1, e2) -> e1.getThread().equals(e2.getThread())
+                            && e1.getGlobalId() < e2.getGlobalId()
+                            && !exec.areMutuallyExclusive(e1, e2));
+            time(definition, start, System.currentTimeMillis());
```
[the remark this answers is not in the document]
Comment (ThomasHaas) [reply dartagnan/src/main/java/com/dat3m/dartagnan/wmm/analysis/LazyRelationAnalysis.java:196]: In this particular example, there is probably not much gain. But in general, it is a good idea to have domain/range prefiltered like this, because you only store linear amount of information to avoid possibly quadratic amount of iteration. That's why the lazy graphs are defined using approximations of domain/range.

File: dartagnan/src/main/java/com/dat3m/dartagnan/encoding/WmmEncoder.java [hunk 5]
```diff
@@ -143,12 +142,77 @@ public BooleanFormula encodeConsistency() {
...
+        List<Axiom> axioms = context.getTask().getMemoryModel().getAxioms();
+        Map<Relation, MutableEventGraph> mutableSets = new HashMap<>();
+        LazyEncodeSets visitor = new LazyEncodeSets(lra, mutableSets);
+        axioms.forEach(a -> a.getEncodeGraph(context.getTask(), context.getAnalysisContext())
+                .forEach((r, eg) -> {
+                    MutableEventGraph copy = new MapEventGraph(eg.getOutMap());
+                    copy.retainAll(lra.getKnowledge(r).getMaySet());
+                    visitor.add(r, copy);
+                    // Force adding must edges to match the result of native analysis
+                    // TODO: Is it really necessary? Method getEventGraph appends them anyway
```
[the remark this answers is not in the document]
Comment (ThomasHaas) [reply dartagnan/src/main/java/com/dat3m/dartagnan/encoding/WmmEncoder.java:211]: This comes from an ambiguity in our modelling. In reality, you have "active sets" which refers to sets of clauses/constraints that need to get respected (i.e., they are not trivially satisfiable) and "relevant literals" which can affect the truth-value of an active constraint.
In the mathematically precise sense, the axioms generate some encoding of arbitrary shape that refers to variables (edges) and all those variables are "relevant". For each relevant variable, its defining constraint now becomes active, which in turn causes other variables to become relevant and this propagates recursively downwards.
Our current modelling uses "encode sets" to kinda refer to both "relevant literals" and "active constraints" though most lymeans the latter (i.e, a `Map<Definition, EventGraph>` would be more correct). And for active constraints, you can actually have an active defining constraint for a variable that is statically known (i.e., inside must or outside may) and that is why "encodeSet" does not always satisfy the expected `must-set \subseteq encodeSet \subseteq may-set` inequality.

That being said, the method `axiom.getEncodeGraph` actually refers to "relevant variables" of the axiom (I know, this is confusing) and there we can indeed (I believe) make sure to never include statically known variables.
Due to the ambiguity, we have thus far allowed `axiom.getEncodeGraph` to contain statically known variables though.

In short, you can remove must-edges from the encode set of axioms if you want.

File: dartagnan/src/main/java/com/dat3m/dartagnan/wmm/utils/graph/immutable/ImmutableMapEventGraph.java [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+    }
+
+    private ImmutableMapEventGraph(Map<Event, Set<Event>> data, int size) {
+        this.data = data.entrySet().stream()
+                .filter(e -> !e.getValue().isEmpty())
+                .collect(ImmutableMap.toImmutableMap(
+                        Map.Entry::getKey,
+                        entry -> entry.getValue() instanceof ImmutableSet
+                                ? entry.getValue()
+                                : ImmutableSet.copyOf(entry.getValue())));
```
Comment (ThomasHaas) [inline dartagnan/src/main/java/com/dat3m/dartagnan/wmm/utils/graph/immutable/ImmutableMapEventGraph.java:33]: `ImmutableSet.copyOf()` already recognizes if it copies another immutable set and then skips the copy.

File: dartagnan/src/main/java/com/dat3m/dartagnan/wmm/utils/graph/immutable/LazyEventGraph.java [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+    public boolean isEmpty() {
+        return empty;
+    }
+
+    @Override
+    public int size() {
+        if (size == -1) {
+            size = domain.stream()
+                    .map(e1 -> range.stream().filter(e2 -> function.test(e1, e2)).count())
+                    .map(s -> Integer.parseInt(s.toString()))
```
Comment (ThomasHaas) [inline dartagnan/src/main/java/com/dat3m/dartagnan/wmm/utils/graph/immutable/LazyEventGraph.java:44]: Why the odd roundtrip through the string representation? Is it just to capture cases where count() gets larger than an int? If so, you can use `Math.toIntExact()`.

**Review by ThomasHaas — COMMENTED** [review]
I saw that you make use of a lot of parallel operations and `ConcurrentHashMap`. Do you by chance have any data/estimate (even anecdotal is fine) on how much this improves performance compared to the standard sequential operations?

**Comment by ThomasHaas** [comment]
Ok, that sounds great. I guess the speedup can be almost linear in the number of available/used threads when doing bulk copying.

File: dartagnan/src/main/java/com/dat3m/dartagnan/wmm/analysis/LazyRelationAnalysis.java [hunk 8]
```diff
@@ -0,0 +1,651 @@
...
+            return new RelationAnalysis.Knowledge(must, must);
+        }
+
+        @Override
+        public RelationAnalysis.Knowledge visitExternal(External definition) {
+            long start = System.currentTimeMillis();
+            EventGraph must = new LazyEventGraph(visibleEvents, visibleEvents,
+                    (e1, e2) -> !e1.getThread().equals(e2.getThread()));
```
[the remark this answers is not in the document]
Comment (hernanponcedeleon) [reply dartagnan/src/main/java/com/dat3m/dartagnan/wmm/analysis/LazyRelationAnalysis.java:173]: Makes sense. I noticed the "inconsistency" with the pattern in other visitors without really considering that this was about `ext` where mutual exclusion does not make much sense.

File: dartagnan/src/main/java/com/dat3m/dartagnan/wmm/analysis/LazyRelationAnalysis.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (hernanponcedeleon) [reply dartagnan/src/main/java/com/dat3m/dartagnan/wmm/analysis/LazyRelationAnalysis.java:228]: Please copy the same FIXME we have in `NativeRelationAnalysis`

File: dartagnan/src/main/java/com/dat3m/dartagnan/encoding/WmmEncoder.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (hernanponcedeleon) [reply dartagnan/src/main/java/com/dat3m/dartagnan/encoding/WmmEncoder.java:211]: @natgavrilenko can you please remove those both here and in the native analysis?

[the remark this answers is not in the document]
Comment (hernanponcedeleon) [reply dartagnan/src/main/java/com/dat3m/dartagnan/encoding/WmmEncoder.java:211]: But then this commit should not add unnecessary code (as suggested by the comment in the commit)

[the remark this answers is not in the document]
Comment (hernanponcedeleon) [reply dartagnan/src/main/java/com/dat3m/dartagnan/encoding/WmmEncoder.java:211]: Then the claim that this commit has nothing to do with the native analysis does not hold and this needs to be fixed.

[the remark this answers is not in the document]
Comment (ThomasHaas) [reply dartagnan/src/main/java/com/dat3m/dartagnan/encoding/WmmEncoder.java:211]: What needs to be fixed? Removing the must-sets also in the native analysis? I think it is fair to not touch the native analysis in this PR, take it as ground truth, and let the lazy analysis mimic its results (as it is done now).

[the remark this answers is not in the document]
Comment (hernanponcedeleon) [reply dartagnan/src/main/java/com/dat3m/dartagnan/encoding/WmmEncoder.java:211]: I think it is a bad idea adding useless stuff to a new implementation just to mimic the behavior of another fishy implementation. But if you both agree that it it's not worth to properly fix this now, I will not object further.
